command/git_push.go

```diff
@@ -0,0 +1,153 @@
+package command
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/evergreen/rest/client"
+	restModel "github.com/evergreen-ci/evergreen/rest/model"
+	"github.com/mitchellh/mapstructure"
+	"github.com/mongodb/grip/level"
+	"github.com/pkg/errors"
+)
+
+// gitMerge is a command that commits tracked changes and pushes the resulting commit to a remote repository
+type gitPush struct {
+	// The root directory (locally) that the code is checked out into.
+	// Must be a valid non-blank directory name.
+	Directory string `plugin:"expand"`
+	base
+}
+
+func gitPushFactory() Command   { return &gitPush{} }
+func (c *gitPush) Name() string { return "git.push" }
+
+// ParseParams parses the command's configuration.
+// Fulfills the Command interface.
+func (c *gitPush) ParseParams(params map[string]interface{}) error {
+	err := mapstructure.Decode(params, c)
+	if err != nil {
+		return err
+	}
+
+	if c.Directory == "" {
+		return errors.Errorf("error parsing '%v' params: value for directory "+
+			"must not be blank", c.Name())
+	}
+
+	return nil
+}
+
+// Execute gets the source code required by the project
+func (c *gitPush) Execute(ctx context.Context, comm client.Communicator, logger client.LoggerProducer, conf *model.TaskConfig) error {
+	p, err := comm.GetTaskPatch(ctx, client.TaskData{ID: conf.Task.Id, Secret: conf.Task.Secret})
+	if err != nil {
+		return errors.Wrap(err, "Failed to get patch")
+	}
+
+	commands := getCheckoutBranchCommands(conf.ProjectRef.Branch)
+	if err = c.runCommands(ctx, commands, c.Directory, logger); err != nil {
+		return errors.Wrapf(err, "can't checkout '%s' branch", conf.ProjectRef.Branch)
+	}
+
+	// fail the merge if HEAD has moved
+	headSHA, err := c.revParse(ctx, logger, "HEAD")
+	if err != nil {
+		return errors.Wrap(err, "can't get SHA for HEAD")
+	}
+	if p.Githash != headSHA {
```

Comment:
I'm not familiar with the use case for this, but would it make sense to try a rebase and still push if no conflicts?

Reply:
The thinking of this check is for this case: 

1. a patch fails and gets dequeued. 
2. the commit queue moves on and merges a bunch more branches.
3. the failed patch is restarted and passes.

Without this check the commit queue would merge the patch now, even though some or all of the changes haven't been tested on top of the current HEAD.
This solution is to record the current tip of the project's branch in the patch as its githash when the commit queue creates the patch and check it when we go to merge.